File: compiler/packages/babel-plugin-react-compiler/src/Validation/ValidateContextVariableLValues.ts
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {CompilerDiagnostic, CompilerError} from '..';
import {ErrorCategory} from '../CompilerError';
import {Environment} from '../HIR/Environment';
import {HIRFunction, IdentifierId, Place} from '../HIR';
import {printPlace} from '../HIR/PrintHIR';
import {eachInstructionValueLValue, eachPatternOperand} from '../HIR/visitors';

/**
 * Validates that all store/load references to a given named identifier align with the
 * "kind" of that variable (normal variable or context variable). For example, a context
 * variable may not be loaded/stored with regular StoreLocal/LoadLocal/Destructure instructions.
 */
export function validateContextVariableLValues(fn: HIRFunction): void {
  const identifierKinds: IdentifierKinds = new Map();
  validateContextVariableLValuesImpl(fn, identifierKinds, fn.env);
}

function validateContextVariableLValuesImpl(
  fn: HIRFunction,
  identifierKinds: IdentifierKinds,
  env: Environment,
): void {
  for (const [, block] of fn.body.blocks) {
    for (const instr of block.instructions) {
      const {value} = instr;
      switch (value.kind) {
        case 'DeclareContext':
        case 'StoreContext': {
          visit(identifierKinds, value.lvalue.place, 'context', env);
          break;
        }
        case 'LoadContext': {
          visit(identifierKinds, value.place, 'context', env);
          break;
        }
        case 'StoreLocal':
        case 'DeclareLocal': {
          visit(identifierKinds, value.lvalue.place, 'local', env);
          break;
        }
        case 'LoadLocal': {
          visit(identifierKinds, value.place, 'local', env);
          break;
        }
        case 'PostfixUpdate':
        case 'PrefixUpdate': {
          visit(identifierKinds, value.lvalue, 'local', env);
          break;
        }
        case 'Destructure': {
          for (const lvalue of eachPatternOperand(value.lvalue.pattern)) {
            visit(identifierKinds, lvalue, 'destructure', env);
          }
          break;
        }
        case 'ObjectMethod':
        case 'FunctionExpression': {
          validateContextVariableLValuesImpl(
            value.loweredFunc.func,
            identifierKinds,
            env,
          );
          break;
        }
        default: {
          for (const _ of eachInstructionValueLValue(value)) {
            fn.env.recordError(
              CompilerDiagnostic.create({
                category: ErrorCategory.Todo,
                reason:
                  'ValidateContextVariableLValues: unhandled instruction variant',
                description: `Handle '${value.kind} lvalues`,
              }).withDetails({
                kind: 'error',
                loc: value.loc,
                message: null,
              }),
            );
          }
        }
      }
    }
  }
}

type IdentifierKinds = Map<
  IdentifierId,
  {place: Place; kind: 'local' | 'context' | 'destructure'}
>;

function visit(
  identifiers: IdentifierKinds,
  place: Place,
  kind: 'local' | 'context' | 'destructure',
  env: Environment,
): void {
  const prev = identifiers.get(place.identifier.id);
  if (prev !== undefined) {
    const wasContext = prev.kind === 'context';
    const isContext = kind === 'context';
    if (wasContext !== isContext) {
      if (prev.kind === 'destructure' || kind === 'destructure') {
        env.recordError(
          CompilerDiagnostic.create({
            category: ErrorCategory.Todo,
            reason: `Support destructuring of context variables`,
            description: null,
          }).withDetails({
            kind: 'error',
            loc: kind === 'destructure' ? place.loc : prev.place.loc,
            message: null,
          }),
        );
        return;
      }

      CompilerError.invariant(false, {
        reason:
          'Expected all references to a variable to be consistently local or context references',
        description: `Identifier ${printPlace(place)} is referenced as a ${kind} variable, but was previously referenced as a ${prev.kind} variable`,
        message: `this is ${prev.kind}`,
        loc: place.loc,
      });
    }
  }
  identifiers.set(place.identifier.id, {place, kind});
}

File: compiler/crates/react_compiler_validation/src/validate_context_variable_lvalues.rs
use std::collections::HashMap;

use react_compiler_diagnostics::{
    CompilerDiagnostic, CompilerDiagnosticDetail, CompilerError, ErrorCategory,
};
use react_compiler_hir::{
    FunctionId, HirFunction, Identifier, IdentifierId, InstructionValue,
    Place,
};
use react_compiler_hir::environment::Environment;
use react_compiler_hir::visitors::{each_instruction_value_lvalue, each_pattern_operand};

/// Variable reference kind: local, context, or destructure.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum VarRefKind {
    Local,
    Context,
    Destructure,
}

impl std::fmt::Display for VarRefKind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            VarRefKind::Local => write!(f, "local"),
            VarRefKind::Context => write!(f, "context"),
            VarRefKind::Destructure => write!(f, "destructure"),
        }
    }
}

type IdentifierKinds = HashMap<IdentifierId, (Place, VarRefKind)>;

/// Validates that context variable lvalues are used consistently.
///
/// Port of ValidateContextVariableLValues.ts
pub fn validate_context_variable_lvalues(
    func: &HirFunction,
    env: &mut Environment,
) -> Result<(), CompilerDiagnostic> {
    validate_context_variable_lvalues_with_errors(func, &env.functions, &env.identifiers, &mut env.errors)
}

/// Like [`validate_context_variable_lvalues`], but writes diagnostics into the
/// provided `errors` instead of `env.errors`. Useful when the caller wants to
/// discard the diagnostics (e.g. when lowering is incomplete).
pub fn validate_context_variable_lvalues_with_errors(
    func: &HirFunction,
    functions: &[HirFunction],
    identifiers: &[Identifier],
    errors: &mut CompilerError,
) -> Result<(), CompilerDiagnostic> {
    let mut identifier_kinds: IdentifierKinds = HashMap::new();
    validate_context_variable_lvalues_impl(func, &mut identifier_kinds, functions, identifiers, errors)
}

fn validate_context_variable_lvalues_impl(
    func: &HirFunction,
    identifier_kinds: &mut IdentifierKinds,
    functions: &[HirFunction],
    identifiers: &[Identifier],
    errors: &mut CompilerError,
) -> Result<(), CompilerDiagnostic> {
    let mut inner_function_ids: Vec<FunctionId> = Vec::new();

    for (_block_id, block) in &func.body.blocks {
        for &instr_id in &block.instructions {
            let instr = &func.instructions[instr_id.0 as usize];
            let value = &instr.value;

            match value {
                InstructionValue::DeclareContext { lvalue, .. }
                | InstructionValue::StoreContext { lvalue, .. } => {
                    visit(identifier_kinds, &lvalue.place, VarRefKind::Context, identifiers, errors)?;
                }
                InstructionValue::LoadContext { place, .. } => {
                    visit(identifier_kinds, place, VarRefKind::Context, identifiers, errors)?;
                }
                InstructionValue::StoreLocal { lvalue, .. }
                | InstructionValue::DeclareLocal { lvalue, .. } => {
                    visit(identifier_kinds, &lvalue.place, VarRefKind::Local, identifiers, errors)?;
                }
                InstructionValue::LoadLocal { place, .. } => {
                    visit(identifier_kinds, place, VarRefKind::Local, identifiers, errors)?;
                }
                InstructionValue::PostfixUpdate { lvalue, .. }
                | InstructionValue::PrefixUpdate { lvalue, .. } => {
                    visit(identifier_kinds, lvalue, VarRefKind::Local, identifiers, errors)?;
                }
                InstructionValue::Destructure { lvalue, .. } => {
                    for place in each_pattern_operand(&lvalue.pattern) {
                        visit(identifier_kinds, &place, VarRefKind::Destructure, identifiers, errors)?;
                    }
                }
                InstructionValue::FunctionExpression { lowered_func, .. }
                | InstructionValue::ObjectMethod { lowered_func, .. } => {
                    inner_function_ids.push(lowered_func.func);
                }
                _ => {
                    for _ in each_instruction_value_lvalue(value) {
                        errors.push_diagnostic(
                            CompilerDiagnostic::new(
                                ErrorCategory::Todo,
                                "ValidateContextVariableLValues: unhandled instruction variant",
                                None,
                            )
                            .with_detail(CompilerDiagnosticDetail::Error {
                                loc: value.loc().copied(),
                                message: None,
                            }),
                        );
                    }
                }
            }
        }
    }

    // Process inner functions after the block loop to avoid borrow conflicts
    for func_id in inner_function_ids {
        let inner_func = &functions[func_id.0 as usize];
        validate_context_variable_lvalues_impl(inner_func, identifier_kinds, functions, identifiers, errors)?;
    }

    Ok(())
}

/// Format a place like TS `printPlace()`: `<effect> <name>$<id>`
fn format_place(place: &Place, identifiers: &[Identifier]) -> String {
    let id = place.identifier;
    let ident = &identifiers[id.0 as usize];
    let name = match &ident.name {
        Some(n) => n.value().to_string(),
        None => String::new(),
    };
    format!("{} {}${}", place.effect, name, id.0)
}

fn visit(
    identifiers: &mut IdentifierKinds,
    place: &Place,
    kind: VarRefKind,
    env_identifiers: &[Identifier],
    errors: &mut CompilerError,
) -> Result<(), CompilerDiagnostic> {
    if let Some((prev_place, prev_kind)) = identifiers.get(&place.identifier) {
        let was_context = *prev_kind == VarRefKind::Context;
        let is_context = kind == VarRefKind::Context;
        if was_context != is_context {
            if *prev_kind == VarRefKind::Destructure || kind == VarRefKind::Destructure {
                let loc = if kind == VarRefKind::Destructure {
                    place.loc
                } else {
                    prev_place.loc
                };
                errors.push_diagnostic(
                    CompilerDiagnostic::new(
                        ErrorCategory::Todo,
                        "Support destructuring of context variables",
                        None,
                    )
                    .with_detail(CompilerDiagnosticDetail::Error {
                        loc,
                        message: None,
                    }),
                );
                return Ok(());
            }
            let place_str = format_place(place, env_identifiers);
            return Err(CompilerDiagnostic::new(
                ErrorCategory::Invariant,
                "Expected all references to a variable to be consistently local or context references",
                Some(format!(
                    "Identifier {} is referenced as a {} variable, but was previously referenced as a {} variable",
                    place_str, kind, prev_kind
                )),
            )
            .with_detail(CompilerDiagnosticDetail::Error {
                loc: place.loc,
                message: Some(format!("this is {}", prev_kind)),
            }));
        }
    }
    identifiers.insert(place.identifier, (place.clone(), kind));
    Ok(())
}
